### DISCUSSION TITLE
Future exciting change to TimeSeries() that may potentially break some code (+ very easy fix)

### DISCUSSION THREAD

**Discussion by felixchenier** [kineticstoolkit/kineticstoolkit#202] [Announcements]
In the next release (0.14), the TimeSeries constructor will change slightly in a way that could break your code if you initialized a TimeSeries using a list of arguments. I don't think many people will be affected because such TimeSeries initialization has never been used in examples or tutorials.

More precisely, if your code creates a TimeSeries from scratch (in opposition to reading a file, e.g., ktk.read_c3d):

### Creating an empty TimeSeries

```python
ts = ktk.TimeSeries()
ts.time = some_time_variable
ts.data["some_data_key"] = some_array
```

### Creating a TimeSeries with time and data as keyword arguments

```python
ts = ktk.TimeSeries(time=some_time_variable, data={"some_data_key": some_array})
```

### Creating a TimeSeries with time and data as a list of arguments

```python
# Won't work in version 0.14
ts = ktk.TimeSeries(some_time_variable, {"some_data_key": some_array})
```

If you happen to have such TimeSeries initialization in your code, it is highly recommended to change for an initialization using keyword arguments (which is generally a better coding practice anyway) before updating to the upcoming 0.14 version.

This is related to new feature #199 which will bring a very exciting way to create a TimeSeries directly from many types of variables, by directly casting these variables to TimeSeries. For instance:

```python
# Coming in next release
ts = ktk.TimeSeries(some_array)
ts = ktk.TimeSeries(some_DataFrame)
ts = ktk.TimeSeries(some_list)

# etc.
```

**Comment by MuhammadUtoro** [comment]
Hi Felix,

I want to ask question regarding file-writing. Since, I use OpenSim quite a lot, is there any possibilities to write .trc or .mot file? because if we want to use marker data in OpenSim we need .trc or .mot file or is this outside of the scope of the package?
Thanks again for your packages! it helps me a lot with my python for biomechanical analysis

Regards,
Muhammad

**Reply by felixchenier** [reply]
Hello @MuhammadUtoro 

This is definitely in the scope of the package. I'll write a feature request to include reading/writing mot/trc files in Kinetics Toolkit, maybe as an extension first.

## In the mean time

We started a small exploratory project to perform standard OpenSim operations right from Kinetics Toolkit a few months ago. You can find it here: https://github.com/felixchenier/kineticstoolkit_opensim/blob/master/kineticstoolkit_opensim.py

It's still very much a work in progress, but I'm quite confident that the functions `read_trc_file` and `write_trc_file` do work. You can download the file in your current directory, import it, and use these functions to generate a `trc` from a TimeSeries, or a TimeSeries from a `trc`.

Thanks for your feedback!

**Reply by MuhammadUtoro** [reply]
Hi @felixchenier 

Thank you for your answer! But the link you provided does not show anything.

**Reply by felixchenier** [reply]
@MuhammadUtoro Weird, maybe an access right problem.

I reviewed the different functions this morning and the one I think you need the most is `write_trc`, so that you can analyze your data in OpenSim. The read functions that I have (for reading mot, sto and trc as ktk.TimeSeries) may not be optimal, I need to polish it before releasing. But do you still need it?

Also, I don't have an export function for forces yet, only for markers. But it's now on the issue tracker as a feature request, and I believe this is an important feature to have.

```
def write_trc(filename: str, markers: ktk.TimeSeries):
    """
    Export a TimeSeries or marker positions to OpenSim's TRC file format.

    Parameters
    ----------
    markers
        Markers trajectories.

    filename
        Name of the trc file to create.

    """
    markers = markers.copy()
    markers.fill_missing_samples(0)

    n_markers = len(markers.data)
    n_frames = markers.time.shape[0]
    data_rate = markers.get_sample_rate()
    camera_rate = markers.get_sample_rate()
    units = "m"

    # Open file
    with open(filename, "w") as fid:
        fid.write(f"PathFileType\t4\t(X/Y/Z)\t{filename}\n")
        fid.write(
            "DataRate\tCameraRate\tNumFrames\tNumMarkers\tUnits\t"
            "OrigDataRate\tOrigDataStartFrame\tOrigNumFrames\n"
        )
        fid.write(
            f"{data_rate}\t{camera_rate}\t{n_frames}\t{n_markers}\t"
            f"{units}\t{data_rate}\t1\t{n_frames}\n"
        )

        # Write marker names
        fid.write("Frame#\tTime")
        for key in markers.data:
            fid.write(f"\t{key}\t\t")
        fid.write("\n")

        # Write coordinate names
        fid.write("\t")
        for i, key in enumerate(markers.data):
            fid.write(f"\tX{i+1}\tY{i+1}\tZ{i+1}")
        fid.write("\n\n")

        # Write trajectories
        for i_frame in range(n_frames):
            fid.write(f"{i_frame+1}\t" "{:.3f}".format(markers.time[i_frame]))
            for key in markers.data:
                fid.write(
                    "\t{:.5f}".format(markers.data[key][i_frame, 0])
                    + "\t{:.5f}".format(markers.data[key][i_frame, 1])
                    + "\t{:.5f}".format(markers.data[key][i_frame, 2])
                )
            fid.write("\n")

```

**Reply by MuhammadUtoro** [reply]
Hi @felixchenier big thanks for your script. No, I am not actually in rush and that write_trc will do it I think. I am just practicing using OpenSim with my own lab data (in Webinar or tutorial they have prepared the .mot, .trc, etc). I am just curious how to do it from scratch. I just amazed with how you build such a kind of package. I will definitely want to learn or contribute if I could.
Thanks again!
